src/main/java/com/kobylynskyi/graphql/codegen/mapper/GraphqlTypeToJavaTypeMapper.java

```diff
@@ -296,67 +295,68 @@ private static String wrapSuperTypeIntoList(MappingContext mappingContext, Strin
+                    !computedTypeName.startsWith(SCALA_UTIL_LIST) && !computedTypeName.startsWith(JAVA_UTIL_LIST)) {
+                // wrap the type into java.util.Optional (except java list and scala list)
+                computedTypeName = getGenericsString(mappingContext, SCALA_UTIL_OPTIONAL, computedTypeName);
+            } else if (!computedTypeName.startsWith(JAVA_UTIL_LIST)) {
+                // wrap the type into java.util.Optional (except lists)
+                computedTypeName = getGenericsString(mappingContext, JAVA_UTIL_OPTIONAL, computedTypeName);
             }
-        } else if (Boolean.TRUE.equals(mappingContext.getUseOptionalForNullableReturnTypes())) {
-            // wrap the type into java.util.Optional (except lists)
-            if (!namedDefinition.isMandatory() && !javaTypeName.startsWith(JAVA_UTIL_LIST)) {
-                return getGenericsString(mappingContext, JAVA_UTIL_OPTIONAL, javaTypeName);
+        }
+        // scala
+        if (GeneratedLanguage.SCALA.equals(mappingContext.getGeneratedLanguage())) {
+            if (computedTypeName.startsWith(SCALA_UTIL_LIST) &&
+                    Utils.isNotBlank(mappingContext.getApiReturnListType())) {
+                // in case it is query/mutation, return type is list and apiReturnListType is set
+                return computedTypeName.replace(SCALA_UTIL_LIST, mappingContext.getApiReturnListType());
             }
-            // wrap the type into java.util.Optional (except java list and scala list)
-            if (!namedDefinition.isMandatory() && !javaTypeName.startsWith(SCALA_UTIL_LIST) && !javaTypeName.startsWith(JAVA_UTIL_LIST)) {
-                return getGenericsString(mappingContext, SCALA_UTIL_OPTIONAL, javaTypeName);
+            if (Utils.isNotBlank(mappingContext.getApiReturnType())) {
+                // in case it is query/mutation and apiReturnType is set
+                return getGenericsString(mappingContext, mappingContext.getApiReturnType(), computedTypeName);
             }
         } else {
-            // scala
-            if (GeneratedLanguage.SCALA.equals(mappingContext.getGeneratedLanguage())) {
-                if (javaTypeName.startsWith(SCALA_UTIL_LIST) &&
-                        Utils.isNotBlank(mappingContext.getApiReturnListType())) {
-                    // in case it is query/mutation, return type is list and apiReturnListType is set
-                    return javaTypeName.replace(SCALA_UTIL_LIST, mappingContext.getApiReturnListType());
-                }
-                if (Utils.isNotBlank(mappingContext.getApiReturnType())) {
-                    // in case it is query/mutation and apiReturnType is set
-                    return getGenericsString(mappingContext, mappingContext.getApiReturnType(), javaTypeName);
-                }
-            }
-            if (javaTypeName.startsWith(JAVA_UTIL_LIST) &&
+            if (computedTypeName.startsWith(JAVA_UTIL_LIST) &&
```

Comment:
Can  use `switch` ? I'm implementing kotlin, and there's going to be a lot of conflict.

Reply:
@jxnu-liguobin before implementing kotlin we need to refactor the code to make it simple for further extensions. I was thinking to do it later next week. So please don't do kotlin yet :)

Reply:
In fact, the templates and code have been written https://github.com/jxnu-liguobin/graphql-java-codegen/tree/kotlin/src/main/resources/templates/kotlin-lang (In addition to the unit test, the rest are basically ready
), but there is still one problem that has not been solved. I am ready to wait until you refactor.

one problem is due to the strict restrictions of kotlin, overriding the parent class must show adding `override`, which is different from the Scala and Java, which makes it may need record all abstract methods of the parent class.

```kotlin
package com.github.graphql

import kotlin.properties.Delegates
import com.kobylynskyi.graphql.codegen.model.graphql.GraphQLRequestSerializer
import java.util.StringJoiner

@javax.annotation.Generated(
    value = ["com.kobylynskyi.graphql.codegen.GraphQLCodegen"],
    date = "2020-12-31T23:59:59-0500"
)
open class GithubReopenedEventTO(
    @JvmField private var actor: GithubActorTO,
    @javax.validation.constraints.NotNull
    @JvmField private var closable: GithubClosableTO,
    @javax.validation.constraints.NotNull
    @JvmField private var createdAt: String,
    @javax.validation.constraints.NotNull
    @JvmField private var id: String
) : GithubIssueTimelineItemsTO, GithubPullRequestTimelineItemsTO, GithubIssueTimelineItemTO, GithubPullRequestTimelineItemTO, GithubNodeTO {

    fun getActor(): GithubActorTO {
        return actor
    }
    fun setActor(actor: GithubActorTO) {
        this.actor = actor
    }
    fun getClosable(): GithubClosableTO {
        return closable
    }
    fun setClosable(closable: GithubClosableTO) {
        this.closable = closable
    }
    fun getCreatedAt(): String {
        return createdAt
    }
    fun setCreatedAt(createdAt: String) {
        this.createdAt = createdAt
    }
    fun getId(): String { // if super has it, must add `override`,like : override fun getId()
        return id
    }
    fun setId(id: String) {
        this.id = id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as GithubReopenedEventTO
        return Objects.equals(actor, that.actor)
            && Objects.equals(closable, that.closable)
            && Objects.equals(createdAt, that.createdAt)
            && Objects.equals(id, that.id);
    }

    override fun hashCode(): Int {
        return Objects.hash(actor, closable, createdAt, id);
    }

    // In the future, it will change.
    override fun toString(): String {
        val joiner = StringJoiner(", ", "{ ", " }")
        joiner.add("actor: " + GraphQLRequestSerializer.getEntry(actor))
        joiner.add("closable: " + GraphQLRequestSerializer.getEntry(closable))
        joiner.add("createdAt: " + GraphQLRequestSerializer.getEntry(createdAt))
        joiner.add("id: " + GraphQLRequestSerializer.getEntry(id))
        return joiner.toString()
    }

    class Builder {

        private lateinit var actor: GithubActorTO
        private lateinit var closable: GithubClosableTO
        private lateinit var createdAt: String
        private lateinit var id: String

        fun setActor(actor: GithubActorTO): Builder {
            this.actor = actor
            return this
        }

        fun setClosable(closable: GithubClosableTO): Builder {
            this.closable = closable
            return this
        }

        fun setCreatedAt(createdAt: String): Builder {
            this.createdAt = createdAt
            return this
        }

        fun setId(id: String): Builder {
            this.id = id
            return this
        }

        fun build(): GithubReopenedEventTO {
            return GithubReopenedEventTO(actor, closable, createdAt, id)
        }
    }
}
```